pennylane/beta/_qubit_device.py

```diff
@@ -0,0 +1,634 @@
+        can be used by devices in :meth:`apply` for parametric compilation."""
+
+    @classmethod
+    def capabilities(cls):
+
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="qubit",
+            supports_finite_shots=True,
+            supports_tensor_observables=True,
+            returns_probs=True,
+        )
+        return capabilities
+
+    def reset(self):
+        """Reset the backend state.
+
+        After the reset, the backend should be as if it was just constructed.
+        Most importantly the quantum state is reset to its initial value.
+        """
+        self._samples = None
+        self._circuit_hash = None
+
+    def execute(self, circuit, **kwargs):
+        """Execute a queue of quantum operations on the device and then
+        measure the given observables.
+
+        For plugin developers: instead of overwriting this, consider
+        implementing a suitable subset of
+
+        * :meth:`apply`
+
+        * :meth:`~.generate_samples`
+
+        * :meth:`~.probability`
+
+        Additional keyword arguments may be passed to the this method
+        that can be utilised by :meth:`apply`. An example would be passing
+        the ``QNode`` hash that can be used later for parametric compilation.
+
+        Args:
+            circuit (~.CircuitGraph): circuit to execute on the device
+
+        Raises:
+            QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
+
+        Returns:
+            array[float]: measured value(s)
+        """
+        self.check_validity(circuit.operations, circuit.observables)
+
+        self._circuit_hash = circuit.hash
+
+        if circuit.is_batched:
+            # redirect entire execution to other method
+            return self.batch_execute(circuit, **kwargs)
+
+        else:
+            return self.single_execute(circuit, **kwargs)
+
+    def single_execute(self, circuit, **kwargs):
+        """Execute a single circuit."""
+        # apply all circuit operations
+        self.apply(circuit.operations, rotations=circuit.diagonalizing_gates, **kwargs)
+
+        # generate computational basis samples
+        if (not self.analytic) or circuit.is_sampled:
+            self._samples = self.generate_samples()
+
+        # compute the required statistics
+        results = self.statistics(circuit.observables)
+
+        # Ensures that a combination with sample does not put
+        # expvals and vars in superfluous arrays
+        all_sampled = all(obs.return_type is Sample for obs in circuit.observables)
+        if circuit.is_sampled and not all_sampled:
+            return self._asarray(results, dtype="object")
+
+        return self._asarray(results)
+
+    def batch_execute(self, circuit, **kwargs):
+        """Execute a circuit that contains batch tapes."""
+
+        if hasattr(self, "batch_apply"):
```

Comment:
Current idea here is that if a device implements ``batch_apply``, we consider it as a device that can "read" batch circuits.